vm/systemSmartContracts/governance.go

```diff
@@ -147,73 +165,268 @@ func (g *governanceContract) init(args *vmcommon.ContractCallInput) vmcommon.Ret
 	}
-	minPass, okConvert := big.NewInt(0).SetString(string(args.Arguments[3]), conversionBase)
-	if !okConvert || minPass.Cmp(big.NewInt(0)) < 0 {
-		g.eei.AddReturnMessage("changeConfig fourth argument is incorrectly formatted")
+	voteOption, err := g.castVoteType(string(args.Arguments[1]))
+	if err != nil {
+		g.eei.AddReturnMessage(err.Error())
 		return vmcommon.UserError
 	}
 
-	scConfig, err := g.getConfig()
+	votePower := big.NewInt(0).SetBytes(args.Arguments[2])
+	delegatedTo, err := g.getDelegatedToAddress(args)
 	if err != nil {
-		g.eei.AddReturnMessage("changeConfig error " + err.Error())
+		g.eei.AddReturnMessage(err.Error())
+		return vmcommon.UserError
+	}
+	totalVotingPower, err := g.getValidatorVotingPower(voterAddress)
+	if err != nil {
+		g.eei.AddReturnMessage(err.Error())
 		return vmcommon.UserError
 	}
 
-	scConfig.NumNodes = numNodes.Int64()
-	scConfig.MinQuorum = int32(minQuorum.Int64())
-	scConfig.MinVetoThreshold = int32(minVeto.Int64())
-	scConfig.MinPassThreshold = int32(minPass.Int64())
+	currentVoteData, err := g.getOrCreateVoteData(proposalToVote, voterAddress)
+	if err != nil {
+		g.eei.AddReturnMessage(err.Error())
+		return vmcommon.ExecutionFailed
+	}
+	if totalVotingPower.Cmp(big.NewInt(0).Add(votePower, currentVoteData.UsedPower)) == -1 {
+		g.eei.AddReturnMessage("not enough voting power to cast this vote")
+		return vmcommon.UserError
+	}
 
-	marshaledData, err := g.marshalizer.Marshal(scConfig)
+	currentVote := &VoteDetails{
+		Value:       voteOption,
+		Power:       votePower,
+		DelegatedTo: delegatedTo,
+		Balance:     big.NewInt(0),
+		Type:        Validator,
+	}
+	newVoteData, updatedProposal, err := g.applyVote(currentVote, currentVoteData, proposal)
 	if err != nil {
-		g.eei.AddReturnMessage("changeConfig error " + err.Error())
+		g.eei.AddReturnMessage(err.Error())
 		return vmcommon.UserError
 	}
-	g.eei.SetStorage([]byte(governanceConfigKey), marshaledData)
+
+	err = g.saveNewVoteData(voterAddress, newVoteData, updatedProposal, Validator)
+	if err != nil {
+		g.eei.AddReturnMessage(err.Error())
+		return vmcommon.ExecutionFailed
+	}
 
 	return vmcommon.Ok
 }
 
-func (g *governanceContract) getConfig() (*GovernanceConfig, error) {
-	marshaledData := g.eei.GetStorage([]byte(governanceConfigKey))
-	scConfig := &GovernanceConfig{}
-	err := g.marshalizer.Unmarshal(scConfig, marshaledData)
+// claimFunds returns back the used funds for a particular proposal if they are unlocked. Accepts a single parameter:
+//  args.Arguments[0] - proposal reference
+func (g *governanceContract) claimFunds(args *vmcommon.ContractCallInput) vmcommon.ReturnCode {
+	if args.CallValue.Cmp(big.NewInt(0)) != 0 {
+		g.eei.AddReturnMessage("invalid callValue, should be 0")
+		return vmcommon.UserError
+	}
+
+	if len(args.Arguments) != 1 {
+		g.eei.AddReturnMessage("invalid number of arguments, expected 1")
+		return vmcommon.FunctionWrongSignature
+	}
+
+	lock := g.getLock(args.CallerAddr, Account, args.Arguments[0])
```

Comment:
nonce and round data could have been easily saved as uint64. no need for big ints.

Reply:
The lock is saved as plain bytes in the storage right now so it would be easier to retrieve by both staking and governance contract, that's why I used big ints, easy to set bytes and convert back to bytes. We can change if you think it's better to use uints

Reply:
even if it is uint64 si saved as bigInt.SetUint64.Bytes and reverse. but nonce and round should be uint

Reply:
done